WFLY-10057 EJB subsystem configure max threads and core threads independently

chengfang opened the pull request:
This PR adds a `core-threads` attribute to wildfly ejb3 subsystem thread-pool configuration, to support independent configuration of `core-threads` and `max-threads`, proper timeout of non-core threads, and reuse existing threads as much as possible without creating new threads.  The implementation of this feature uses an executor service backed by `org.jboss.threads.EnhancedQueueExecutor`.  This PR requires wildfly-core to be upgraded to incorporate wildfly-core PR:
https://github.com/wildfly/wildfly-core/pull/3880/

JIRA: https://issues.jboss.org/browse/WFLY-10057

Related analysis document proposal:
https://github.com/wildfly/wildfly-proposals/pull/227

Related Issues:  
https://issues.jboss.org/browse/EAP7-488
Fine-grained control on ejb3 thread-pool and reuse idle threads in pool before creating new

https://issues.jboss.org/browse/WFCORE-1446
EJB thread pool keepalive not honored and is not reusing inactive threads

chengfang committed:
WFLY-10057 EJB subsystem configure max threads and core threads independently

chengfang committed:
change default thread-pool keepalive-time from 100 milliseconds to 60 seconds.

chengfang committed:
use dynamically-named capability when creating ejb3 thread pool resource; rearranged some imports.

kabir reviewed (changes requested):
@chengfang We need some transformer tests for the added attribute

chengfang committed:
Add to transformer tests for new attribute core-threads in ejb3 thread-pool.

jamezp commented on ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemXMLPersister.java:
```diff
@@ -22,6 +22,11 @@
 
 package org.jboss.as.ejb3.subsystem;
 
+import static org.jboss.as.ejb3.subsystem.EJB3SubsystemModel.*;
```
This is banned by checkstyle. We can't have wildcard imports.

mnovak committed:
Add tests for EAP7-488 / WFLY-10057

chengfang committed:
Merge pull request #1 from mnovak1/WFLY-10057-thread-pool

 Add tests for EAP7-488 / WFLY-10057

chengfang committed:
Add community documentation for the new attribute core-threads in ejb3 thread-pool configuration, and its related changes; changed wildcard static imports to specific imports in EJB3SubsystemXMLPersister.

chengfang commented:
> @chengfang We need some transformer tests for the added attribute

@kabir I've added community documentation for this issue to docs/.../EJB3.adoc, and also added transformer tests for the new attribute.

bstansberry commented:
retest this please

mnovak committed:
Fix Ejb3NonCoreThreadTimeoutTestCase and Ejb3ThreadReuseTestCase test with security manager

mnovak1 commented:
@chengfang thanks for spotting this, here is the fix to test - could you merge it:
https://github.com/chengfang/wildfly/pull/2

mnovak1 commented:
retest this please

chengfang committed:
Merge pull request #2 from mnovak1/WFLY-10057-thread-pool

Fix Ejb3NonCoreThreadTimeoutTestCase and Ejb3ThreadReuseTestCase test…

kabir reviewed (commented):
@chengfang I've only taken a super-quick look and I think the transformer test looks ok. However, I notice that the non-rejecting transformer test using subsystem-ejb3-transform.xml does not contain a `<thread-pool>` entry. Would it be possible to add that (without the new `<core-threads>` element)? That way we get coverage that the subsystem can work with an older slave.

chengfang commented:
> the non-rejecting transformer test using subsystem-ejb3-transform.xml does not contain a entry. Would it be possible to add that (without the new element)?

@kabir can you be more specific?  For `subsystem-ejb3-transform.xml`, I merely updated the schema version from `5.0` to `6.0` to use the current version.  Did you mean to add more xml elements and attributes (that are not new) to `subsystem-ejb3-transform.xml` so that the non-rejecting tests will cover more?

kabir commented:
@chengfang Yes, exactly - it is the 'missing' <thread-pool> that stands out since that is what has the new core-threads element/attribute

chengfang commented:
> it is the 'missing' `<thread-pool>` that stands out since that is what has the new core-threads element/attribute

@kabir `subsystem-ejb3-transform.xml` already contains a `thread-pool` element, around line 47. Any other entries are missing in `subsystem-ejb3-transform.xml`?

kabir commented:
@chengfang you are right, sorry for the noise. I will take a better look in the morning

kabir commented:
This should be good to go once the retest I just triggered on TC has finished
